packages/nimble-components/src/table-column/number-text/models/tests/decimal-formatter.spec.ts

```diff
@@ -125,33 +128,47 @@ describe('DecimalFormatter', () => {
         }
     ] as const;
 
-    const focused: string[] = [];
-    const disabled: string[] = [];
     for (const locale of locales) {
-        for (const testCase of testCases) {
-            const specType = getSpecTypeByNamedList(
-                testCase,
-                focused,
-                disabled
-            );
+        parameterizeNamedList(testCases, (spec, name, value) => {
             // eslint-disable-next-line @typescript-eslint/no-loop-func
-            specType(
-                `${testCase.name} with '${locale}' locale`,
-                // eslint-disable-next-line @typescript-eslint/no-loop-func
-                () => {
-                    const formatter = new DecimalFormatter(
-                        locale,
-                        testCase.minDigits,
-                        testCase.maxDigits
-                    );
-                    const formattedValue = formatter.formatValue(
-                        testCase.value
-                    );
-                    expect(formattedValue).toEqual(
-                        testCase.expectedFormattedValue[locale]
-                    );
-                }
-            );
-        }
+            spec(`${name} with '${locale}' locale`, () => {
+                const formatter = new DecimalFormatter(
+                    locale,
+                    EmptyUnitScaleFormatter,
+                    value.minDigits,
+                    value.maxDigits
+                );
+                const formattedValue = formatter.formatValue(value.value);
+                expect(formattedValue).toEqual(
+                    value.expectedFormattedValue[locale]
+                );
+            });
+        });
     }
+
+    describe('with unit', () => {
+        class TestUnitScaleFormatter extends UnitScaleFormatter {
+            public override getSupportedScaledUnits(): ScaledUnit[] {
+                return [1, 2, 4].map(scaleFactor => {
```

Comment:
I feel like I would expect a whole lot more test cases similar to the decimal formatter test cases, particularly around rounding scenarios + scaleFactor boundaries

For example, 999.999 with 0 decimal digits will round up to 1000, on metric uunits will this render as 1K?

Reply:
All the other test cases above for the decimal formatter validate interaction with the `UnitScaleFormatter` via the `EmptyUnitScaleFormatter`. This covers most logic, except that the `EmptyUnitScaleFormatter` only has a single scaled unit, so we need additional tests to cover the interaction of unit selection with rounding. I think that's the only thing that was missing.

Good catch though, because we weren't handling that case properly! I've added logic to `UnitScaleFormatter` to handle rounding, and associated test cases to `unit-sale-formatter.spec.ts`.

Reply:
I'm skeptical about the algorithm put in.. I'm concerned that:

1. We are picking out one specific setting from formatterOptions to handle but our configuration allows arbitrary numeric formatter settings through. Can any of those other settings impact the rounding behavior? Like significant digits on exponentials might cause rounding right 1255 with two significant digits is 1.3e3.
2. Does this algorithm actually work for arbitrary scales? Like at the KiB to MiB boundary like 1,048.5 KiB with 0 decimals should round up to 1MiB due to the 1,048,576 byte boundary

Reply:
1. You're right. 999 bytes with `maximumSignificantDigits=2` and `notation='scientific'` would be formatted as `1E3 bytes` rather than `1E0 kB` as it should. Moreover, there are other formatter options (including `roundingMode`, `roundingPriority`, and `roundingIncrement`) that affect the rounding behavior. The only approach I can think of to address this is to call `formatter.format()` for each unit, parse the number out of the resulting string, then test if it is >=1. I'm afraid that formatting and parsing O(n) times for every cell (even if n is a dozen or less) may be too inefficient. I should probably profile this. Another approach could be to put guardrails on our implementation by throwing errors if certain formatting options are passed into the `UnitScaleFormatter` constructor. What are your thoughts?
2. Yes, it works. Rounding behavior is unaffected by the conversion factor between units. So for example, if we start with 1,048,500 bytes, we test whether that can be represented as MiB by doing `1,048,500 / 1024^2 = 0.9999275`, then we see that rounds up to 1, so we pick MiB as the unit. The tests I have for this in `unit-scale-formatter.spec.ts` actually use a test formatter with non-uniform scaling factors 0.5, 1, 5, and 10.

Reply:
>  The only approach I can think of to address this is to call formatter.format() for each unit, parse the number out of the resulting string, then test if it is >=1. I'm afraid that formatting and parsing O(n) times for every cell (even if n is a dozen or less) may be too inefficient. I should probably profile this. 

I don't think I follow what algorithm is being proposed that results in `formatting and parsing O(n) times for every cell`. My first thought is it would be 1 or 2 times? (if it matches the current scale or the next up scale) Or maybe it's a worst case O(n) but average is 1 or 2?

> Another approach could be to put guardrails on our implementation by throwing errors if certain formatting options are passed into the UnitScaleFormatter constructor.

Sounds like advocating for a separate column at that point where only supported options are exposed. I'd probably not lean back in that direction.

> The tests I have for this in unit-scale-formatter.spec.ts actually use a test formatter with non-uniform scaling factors 0.5, 1, 5, and 10.

I'd appreciate some additional tests that more directly show real world examples. The invented scaling is probably useful but really hard to reason about.

Reply:
See my [response to an earlier thread about performance](https://github.com/ni/nimble/pull/1620#discussion_r1388383086). There are contrived scenarios where rounding do more than just picking the next unit up. I've implemented an algorithm (in `UnitScaleFormatter.format()`) that generally will result in two calls to `Intl.NumberFormat.formatToParts()`. In the worst case, it may call it N times, where N is the number of possible scaled units, though that would require configuration that we will almost surely never use.

> I'd appreciate some additional tests that more directly show real world examples (behaviors of real world scales).

I changed the `UnitScaleFormatter` tests so that most tests use byte scaled units, and only a few cases that can't be exercised by a byte scale are using my invented scaling.